### PR TITLE
add storage node voucher request service

### PR DESCRIPTION
**Pull request by cam-a**
What: Add a service on the storage node to request vouchers from satellites and insert them into a DB table

Why: For kademlia audit gating. To prevent sybil attacks
https://storjlabs.atlassian.net/browse/V3-1868

(Discussion)
- How many trusted satellites do we expect a storage node to potentially have in the future? This may impact how we want to design db queries

- Right now, the voucher service iterates through the trusted satellites list and calls `NeedVoucher` on each satellite, which determines if a voucher is needed based on whether a voucher exists for this satellite and if that voucher is going to expire. With this method, the number of queries equals the number of satellites in the trusted list (could be thousands in the future?)

- In order to cut down on queries, we could query the vouchersDB for all expiring vouchers, but we would also need to determine which satellites in the trusted list are not in the vouchersDB yet, which would require making another query to return all satellites in the vouchersDB and filter them from the trusted satellites list

- Another idea is to only make one query, which returns all satellite ids and expirations from the table. Since in the idea above we need to return all satellite ids in the table, why not just return all satellite ids and expirations with one query? However, this could be quite large. cpu resources vs db resources?

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [x] Does the PR describe what changes are being made?
 - [x] Does the PR describe why the changes are being made?
 - [x] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [x] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [x] Could the PR be broken into smaller PRs?
 - [x] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [x] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [x] Does any documentation need updating?

### CODE REVIEW

File: internal/testplanet/planet.go [hunk 1]
```diff
@@ -51,13 +51,14 @@ import (
...
 	"storj.io/storj/satellite/satellitedb"
-	"storj.io/storj/satellite/vouchers"
+	satVouchers "storj.io/storj/satellite/vouchers"
 	"storj.io/storj/storagenode"
 	"storj.io/storj/storagenode/collector"
 	"storj.io/storj/storagenode/monitor"
 	"storj.io/storj/storagenode/orders"
 	"storj.io/storj/storagenode/piecestore"
 	"storj.io/storj/storagenode/storagenodedb"
+	"storj.io/storj/storagenode/vouchers"
```
Comment (jenlij) [inline internal/testplanet/planet.go]: can we name this one snVouchers?

File: satellite/vouchers/service.go [hunk 2]
```diff
@@ -49,29 +49,29 @@ func NewService(log *zap.Logger, satellite signing.Signer, cache *overlay.Cache,
...
 
 // Request receives a voucher request and returns a voucher and an error
-func (service *Service) Request(ctx context.Context, req *pb.VoucherRequest) (_ *pb.Voucher, err error) {
+func (service *Service) Request(ctx context.Context, req *pb.VoucherRequest) (_ *pb.VoucherResponse, err error) {
 	defer mon.Task()(&ctx)(&err)
 
 	peer, err := identity.PeerIdentityFromContext(ctx)
 	if err != nil {
-		return &pb.Voucher{}, Error.Wrap(err)
+		return nil, Error.Wrap(err)
```
Comment (jenlij) [inline satellite/vouchers/service.go]: Why are we returning nil here instead of an empty voucher? Just wondering

Comment (cam-a) [reply satellite/vouchers/service.go]: Not sure if there is a particular reason to choose one over the other

File: storagenode/peer.go [hunk 3]
```diff
@@ -110,6 +114,9 @@ type Peer struct {
 		Sender    *orders.Sender
 	}
 
+	//TODO: Where to put this? Storage2/Kademlia?
```
Comment (jenlij) [inline storagenode/peer.go]: What do you mean?

Comment (cam-a) [reply storagenode/peer.go]: Other services have been compartmentalized into the Storage2 and Kademlia structs, but Vouchers don't really fall into either of those so, I could probably get rid of this comment

File: internal/testplanet/planet.go [hunk 4]
```diff
@@ -517,12 +518,13 @@ func (planet *Planet) newSatellites(count int) ([]*satellite.Peer, error) {
 				PasswordCost:    console.TestPasswordCost,
 				AuthTokenSecret: "my-suppa-secret-key",
 			},
+			Vouchers: satVouchers.Config{
```
Comment (jenlij) [inline internal/testplanet/planet.go]: Remember to let devops know about these config changes in the config-changes slack channel

File: satellite/vouchers/service.go (hunk 2 above)
Comment (jenlij) [reply satellite/vouchers/service.go]: Yeah the internet says it doesn't really matter, but then some people say nil is better than a pointer to an empty struct I think because it may be clearer. But in either case, you're fine here!

File: storagenode/peer.go (hunk 3 above)
Comment (jenlij) [reply storagenode/peer.go]: ok yeah even though vouchers are tied to kad, it's kind of its own thing

File: storagenode/storagenodedb/vouchers.go [hunk 5]
```diff
@@ -56,35 +56,27 @@ func (db *vouchersdb) Put(ctx context.Context, voucher *pb.Voucher) (err error)
 	return err
 }
 
-// GetExpiring retrieves all vouchers that are expired or about to expire
-func (db *vouchersdb) GetExpiring(ctx context.Context, expirationBuffer time.Duration) (satellites []storj.NodeID, err error) {
+// NeedVoucher returns true if a satellite's voucher is expired/about to expire or does not exist
```
Comment (jenlij) [inline storagenode/storagenodedb/vouchers.go]: super nit: "NeedVoucher returns true if a voucher from a particular satellite is expired, about to expire, or does not exist"

File: storagenode/vouchers/vouchers_test.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+		expiration, err := ptypes.TimestampProto(time.Now().UTC().Add(24 * time.Hour))
+		require.NoError(t, err)
+
+		voucher := &pb.Voucher{
+			SatelliteId:   satellite.ID,
+			StorageNodeId: storagenode.ID,
+			Expiration:    expiration,
+		}
+
+		// GetValid with no entry
```
Comment (jenlij) [inline storagenode/vouchers/vouchers_test.go]: Could we maybe elaborate more on these test case descriptions? Like "Tests that GetValid returns nil when there is no voucher corresponding to satellite.ID"

File: storagenode/vouchers/vouchers_test.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+		result, err := vdb.GetValid(ctx, []storj.NodeID{satellite.ID})
+		require.NoError(t, err)
+		assert.Nil(t, result)
+
+		// basic Put test
+		err = vdb.Put(ctx, voucher)
+		require.NoError(t, err)
+
+		// basic NeedVoucher test
+		expirationBuffer := 48 * time.Hour
```
Comment (jenlij) [inline storagenode/vouchers/vouchers_test.go]: Why does case result in NeedVoucher true?

**Comment by jenlij** [comment]
Going through the checklist, I realized we should update the design doc and or create new documentation with all the audit gating information once that's all implemented.

File: storagenode/vouchers/service.go [hunk 8]
```diff
@@ -7,16 +7,165 @@ import (
...
+}
+
+// Service is a service for requesting vouchers
+type Service struct {
+	log *zap.Logger
+
+	kademlia  *kademlia.Kademlia
+	transport transport.Client
+
+	vdb   DB
```
Comment (cam-a) [inline storagenode/vouchers/service.go]: call voucherDB

File: storagenode/vouchers/service.go [hunk 9]
```diff
@@ -7,16 +7,165 @@ import (
...
-	GetExpiring(context.Context, time.Duration) ([]storj.NodeID, error)
 	// GetValid returns one valid voucher from the list of approved satellites
 	GetValid(context.Context, []storj.NodeID) (*pb.Voucher, error)
+	// NeedVoucher returns true if a voucher from a particular satellite is expired, about to expire, or does not exist
+	NeedVoucher(context.Context, storj.NodeID, time.Duration) (bool, error)
+}
+
+// Config defines configuration for requesting vouchers.
+type Config struct {
+	Interval         int `help:"number of days between voucher service iterations" default:"7"`
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: if interval is 7 days then that means a new node can never be vetted faster than 7 days. Is that what we want? Seems long to me.

File: storagenode/vouchers/service.go [hunk 10]
```diff
@@ -7,16 +7,165 @@ import (
...
+	ExpirationBuffer int `help:"buffer period of X days into the future. If a voucher would expire within this period, send a request for renewal" default:"7"`
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: we might want to hard code `ExpirationBuffer` to be some value longer than `Interval`

File: storagenode/vouchers/service.go [hunk 11]
```diff
@@ -7,16 +7,165 @@ import (
...
+	defer mon.Task()(&ctx)(&err)
+	return service.Loop.Run(ctx, service.RunOnce)
+}
+
+// RunOnce runs one iteration of the voucher request service
+func (service *Service) RunOnce(ctx context.Context) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	service.log.Info("Checking vouchers")
+
+	satellites := service.trust.GetSatellites((ctx))
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: can we call this `trustedSatellites`?

File: storagenode/vouchers/service.go [hunk 12]
```diff
@@ -7,16 +7,165 @@ import (
...
+
+	if len(satellites) > 0 {
+		var group errgroup.Group
+		ctx, cancel := context.WithTimeout(ctx, time.Hour)
+		defer cancel()
+		for _, sat := range satellites {
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: can we call this `satellite` instead of `sat`?

File: storagenode/vouchers/service.go [hunk 13]
```diff
@@ -7,16 +7,165 @@ import (
...
+			sat := sat
+			need, err := service.vdb.NeedVoucher(ctx, sat, service.expirationBuffer)
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: can we call this `needVoucher`

File: storagenode/vouchers/service.go [hunk 14]
```diff
@@ -7,16 +7,165 @@ import (
...
+				group.Go(func() error {
+					service.Request(ctx, sat)
+					return nil
+				})
+			}
+		}
+		_ = group.Wait() // doesn't return errors
+	} else {
+		service.log.Debug("No voucher requests to send")
+	}
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: can we change this to check `if len(satellites) == 0; service.log.Debug("No voucher requests to send"); return nil` first, then we dont need the rest indented

File: storagenode/vouchers/service.go [hunk 15]
```diff
@@ -7,16 +7,165 @@ import (
...
+
+	resp, err := pb.NewVouchersClient(conn).Request(ctx, &pb.VoucherRequest{})
+	if err != nil {
+		return VoucherError.New("failed to start request: %v", err)
+	}
+
+	// TODO: separate status for disqualified nodes?
+	switch resp.GetStatus() {
+	case pb.VoucherResponse_REJECTED:
+		service.log.Info("Voucher request denied. Node does not meet voucher requirements")
```
Comment (JessicaGreben) [inline storagenode/vouchers/service.go]: is there anyway we can provide more info about why it was rejected? Even if we can pass the specific values, but just generally what is being checked

File: storagenode/vouchers/vouchers_test.go [hunk 16]
```diff
@@ -0,0 +1,204 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
```
Comment (JessicaGreben) [inline storagenode/vouchers/vouchers_test.go]: yay so many tests!

File: storagenode/vouchers/vouchers_test.go [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+
+		// run service and check vouchers have been received
+		err := node.Vouchers.RunOnce(ctx)
+		require.NoError(t, err)
+
+		for _, sat := range planet.Satellites {
+			voucher, err := node.DB.Vouchers().GetValid(ctx, []storj.NodeID{sat.ID()})
+			require.NoError(t, err)
+			assert.NotNil(t, voucher)
+		}
```
Comment (JessicaGreben) [inline storagenode/vouchers/vouchers_test.go]: this is totally unnecessary since your test coverage is already amazing, but just as a fun idea it might be cool to write a test that runs the voucher service when the storagenode is not vetted, then runs it again later and show that it is vetted

File: storagenode/vouchers/vouchers_test.go [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+	})
+}
+
+func TestVerifyVoucher(t *testing.T) {
+	testplanet.Run(t, testplanet.Config{
+		SatelliteCount: 2, StorageNodeCount: 1, UplinkCount: 1,
```
Comment (JessicaGreben) [inline storagenode/vouchers/vouchers_test.go]: do you need uplink?

File: storagenode/vouchers/vouchers_test.go (hunk 7 above)
Comment (cam-a) [reply storagenode/vouchers/vouchers_test.go]: The voucher's expiration, `time.Now().UTC().Add(24 * time.Hour)` falls within the expiration buffer of 48 hours from now

File: storagenode/vouchers/service.go (hunk 9 above)
Comment (jtolds) [reply storagenode/vouchers/service.go]: can we make this be a time.Duration instead of an int plz? the default value (if 7 days) would then be "168h"

File: storagenode/vouchers/service.go (hunk 10 above)
Comment (jtolds) [reply storagenode/vouchers/service.go]: also can this be a time.Duration instead of an int?

File: storagenode/vouchers/service.go (hunk 9 above)
Comment (cam-a) [reply storagenode/vouchers/service.go]: I was torn between time.Duration for code simplicity and int for simplicity of user configuration, but changing it sounds good!

File: storagenode/vouchers/service.go (hunk 10 above)
Comment (cam-a) [reply storagenode/vouchers/service.go]: Jessica and I were thinking it might be better to hard-code this to be == `Interval` or `Interval` + (some extra duration). What do you think?

File: pkg/pb/vouchers.proto [hunk 19]
```diff
@@ -21,6 +21,19 @@ message Voucher {
 
 message VoucherRequest {}
 
+message VoucherResponse {
+    enum Status {
+        INVALID  = 0;
```
Comment (JessicaGreben) [inline pkg/pb/vouchers.proto]: do we ever use this invalid status? seems like it might not be needed

File: pkg/pb/vouchers.proto [hunk 20]
```diff
@@ -21,6 +21,19 @@ message Voucher {
...
+    Voucher voucher = 1;
+
+    Status status = 2;
+}
+
+
 service Vouchers {
-    rpc Request(VoucherRequest) returns (Voucher) {}
+    rpc Request(VoucherRequest) returns (VoucherResponse) {}
 }
```
Comment (JessicaGreben) [inline pkg/pb/vouchers.proto]: add new line at end please

File: storagenode/peer.go [hunk 21]
```diff
@@ -265,6 +271,29 @@ func New(log *zap.Logger, full *identity.FullIdentity, db DB, config Config, ver
...
+		intervalDuration, err := time.ParseDuration(fmt.Sprintf("%dh", intervalHours))
+		if err != nil {
+			return nil, err
+		}
+
+		if config.Vouchers.ExpirationBuffer <= 0 {
+			return nil, errs.New("voucher service expiration buffer (%d) must be > 0", config.Vouchers.ExpirationBuffer)
+		}
+		bufferHours := config.Vouchers.ExpirationBuffer * 24
+		bufferDuration, err := time.ParseDuration(fmt.Sprintf("%dh", bufferHours))
```
Comment (jtolds) [inline storagenode/peer.go]: you don't have to do any of this if you just make the config type be a time.Duration off the bat. our config system knows how to handle time.Durations

File: storagenode/vouchers/service.go (hunk 14 above)
Comment (JessicaGreben) [reply storagenode/vouchers/service.go]: im running storj-sim right now and its not clear what this message means

```
DEBUG	vouchers	vouchers/service.go:107	No voucher requests to send
```
Maybe we could add something like, "No trusted satellites so not requesting any vouchers to approve storing data".  What do you think?

File: pkg/pb/vouchers.proto (hunk 19 above)
Comment (cam-a) [reply pkg/pb/vouchers.proto]: Yeah, we can probably get rid of it

File: internal/testplanet/planet.go [hunk 22]
```diff
@@ -517,12 +518,13 @@ func (planet *Planet) newSatellites(count int) ([]*satellite.Peer, error) {
...
+				Expiration: 30,
```
Comment (jtolds) [inline internal/testplanet/planet.go]: is this a time.Duration? if it is, then this means 30 nanoseconds. if it's not a time.Duration, can we make it one, and then change this to 30*24*time.Day?

File: storagenode/peer.go [hunk 23]
```diff
@@ -265,6 +271,17 @@ func New(log *zap.Logger, full *identity.FullIdentity, db DB, config Config, ver
 		)
 	}
 
+	{ // setup vouchers
+		interval := config.Vouchers.Interval
+		bufferHours := interval.Hours() + 1.0
+		bufferDuration, err := time.ParseDuration(fmt.Sprintf("%fh", bufferHours))
+		if err != nil {
+			return nil, errs.Combine(err, peer.Close())
+		}
```
Comment (jtolds) [inline storagenode/peer.go]: you can just do bufferDuration := interval + time.Hour?

File: internal/testplanet/planet.go (hunk 22 above)
Comment (cam-a) [reply internal/testplanet/planet.go]: ok

File: pkg/pb/vouchers.proto (hunk 19 above)
Comment (jtolds) [reply pkg/pb/vouchers.proto]: it's nice to have the zero value default to "invalid"

Comment (jtolds) [reply pkg/pb/vouchers.proto]: basically, if you construct this protobuf in go, all the values are initialized to zero values, and it's nice in many cases to have the status default to something besides something you opted in to
